File: ios/Detonator/Sources/Detonator/Element/TextElement/TextElement.swift
import UIKit

class TextElement: Element {
    override func decodeAttributes() -> TextAttributes? {
        return super.decodeAttributes()
    }
    
    override func createView() -> TextView {
        return TextView()
    }
    
    override func setUpView() -> Void {
        let view = view as! TextView
        
        view.isUserInteractionEnabled = true
    }
    
    override func patchView() -> Void {
        let view = view as! TextView
        
        let attributes = attributes as! TextAttributes
        let prevAttributes = prevAttributes as! TextAttributes?
        
        let stringBuilder = NSMutableString()
        
        for child in edge!.children {
            if let text = child.text {
                stringBuilder.append(text)
            }
        }
        
        let text = stringBuilder as String
        
        let prevText = view.text
        
        let patchTextBool = forcePatch || text != prevText
        
        if patchTextBool {
            view.text = text
        }
        
        let maxLines = attributes.maxLines
        let prevMaxLines = prevAttributes?.maxLines
        
        let patchMaxLinesBool = forcePatch || maxLines != prevMaxLines
        
        if patchMaxLinesBool {
            let value = maxLines ?? 0
            
            view.numberOfLines = value
        }
    }
    
    override func patchFontSize(fontSize: Float?) -> Void {
        let view = view as! TextView
        
        let value = CGFloat(fontSize ?? 16)
        
        view.font = UIFont.systemFont(ofSize: value)
    }
    
    override func patchColor(color: StyleColor?) -> Void {
        let view = view as! TextView
        
        view.textColor = color != nil ? ColorHelper.parseColor(color: color!) : nil
    }
    
    class TextAttributes: Attributes {
        var maxLines: Int?
        
        required init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            
            maxLines = try container.decodeIfPresent(Int.self, forKey: .maxLines)
            
            try super.init(from: decoder)
        }
        
        private enum CodingKeys: String, CodingKey {
            case maxLines
        }
    }
}

File: android/Detonator/src/main/java/com/iconshot/detonator/element/TextElement.java
package com.iconshot.detonator.element;

import android.util.TypedValue;
import android.widget.TextView;

import com.iconshot.detonator.Detonator;
import com.iconshot.detonator.helpers.ColorHelper;
import com.iconshot.detonator.helpers.CompareHelper;
import com.iconshot.detonator.helpers.ContextHelper;
import com.iconshot.detonator.helpers.PixelHelper;
import com.iconshot.detonator.tree.Edge;

public class TextElement extends Element<TextView, TextElement.Attributes> {
    private int defaultColor;

    public TextElement(Detonator detonator) {
        super(detonator);
    }

    @Override
    protected Class<Attributes> getAttributesClass() {
        return Attributes.class;
    }

    @Override
    protected TextView createView() {
        TextView view = new TextView(ContextHelper.context);

        defaultColor = view.getCurrentTextColor();

        return view;
    }

    @Override
    protected void patchView() {
        if (!edge.children.isEmpty()) {
            StringBuilder stringBuilder = new StringBuilder();

            for (Edge child : edge.children) {
                if (child.text != null) {
                    stringBuilder.append(child.text);
                }
            }

            String text = stringBuilder.toString();

            if (!view.getText().equals(text)) {
                view.setText(text);
            }
        }
    }

    protected void patchFontSize(Float fontSize) {
        float value = ((Integer) PixelHelper.dpToPx(fontSize != null ? fontSize : 16)).floatValue();

        view.setTextSize(TypedValue.COMPLEX_UNIT_PX, value);
    }

    protected void patchColor(Object color) {
        Integer tmpColor = ColorHelper.parseColor(color);

        view.setTextColor(tmpColor != null ? tmpColor : defaultColor);
    }

    protected static class Attributes extends Element.Attributes {}
}
